### PR TITLE
auxdisplay: Add SerLCD auxdisplay driver

### PR DESCRIPTION
**Pull request by janhenke**
SerLCD is an interface for several lcd character display sold by sparkfun.

### CODE REVIEW

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (nordicjm) [inline drivers/auxdisplay/auxdisplay_serlcd.c:159]: Would expect this to do error checking i.e. returns an error if someone tries to move to the 98th column on a display with 12 columns

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (nordicjm) [inline drivers/auxdisplay/auxdisplay_serlcd.c:232]: Return -EINVAL if in invalid entry is provided

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (nordicjm) [inline drivers/auxdisplay/auxdisplay_serlcd.c:291]: *custom

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (nordicjm) [inline drivers/auxdisplay/auxdisplay_serlcd.c:299]: Not seeing any code here that updates the variables keeping a track of the current position

**Review by nordicjm — CHANGES_REQUESTED** [review]

File: drivers/auxdisplay/Kconfig.serlcd
Comment (nordicjm) [inline drivers/auxdisplay/Kconfig.serlcd:5]: I had a search for "SparkFun SerLCD" and found the product page: https://www.sparkfun.com/categories/tags/serlcd which shows LCDs, not VFDs, so if these is not a VFD then please update. Also this suggests that the _is_ a backlight on these but this driver does not support it, are you planning to add support for the backlight in the future?

Comment (janhenke) [reply drivers/auxdisplay/Kconfig.serlcd:5]: Indeed, c&p error.

Regarding the backlight, I did not get the backlight control to work properly yet. So I would like to merge it first without the backlight and add it later, once it works here.

**Comment by janhenke** [comment]
Thank you for your review, I hope I have all comments addressed.

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (nordicjm) [inline drivers/auxdisplay/auxdisplay_serlcd.c:38]: This should be marked as `__packed` and this should also have 2 layouts, one for big endian, one for little endian.

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (nordicjm) [inline drivers/auxdisplay/auxdisplay_serlcd.c:250]: Then shouldn't this be | not &?

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (nordicjm) [inline drivers/auxdisplay/auxdisplay_serlcd.c:289]: Should these be > or >=? 16 columns but for a 0-based index this is 0 to 15

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c:38]: Added packed. I disagree on the 2 layouts, as It's size is only 1 byte, for which the endianess does not make any difference.

**Comment by janhenke** [comment]
Addressed the review comments.

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (nordicjm) [reply drivers/auxdisplay/auxdisplay_serlcd.c]: Actually for a bitfield, it does. Here is a modified hello_world sample:
```
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>

union auxdisplay_display_state {
        uint8_t byte;
        struct __attribute__((__packed__)) {
                uint8_t padding: 5;
                bool power: 1;
                bool cursor: 1;
                bool blinking: 1;
        };
};

union auxdisplay_display_state obj;

int main(void)
{
obj.power = 1;
obj.cursor = 1;
        printf("Hello World! %s\nValue: %x\n", CONFIG_BOARD, obj.byte);
        return 0;
}
```
Here is it running on a little endian ARM cortex:
```
*** Booting Zephyr OS build v3.5.0-rc1-38-gc6c97002b529 ***
Hello World! qemu_cortex_m3
Value: 60
```
And here is it running on a big endian sparc:
```
[93/94] To exit from QEMU enter: 'CTRL+a, x'[QEMU] CPU: leon3
*** Booting Zephyr OS build v3.5.0-rc1-38-gc6c97002b529 ***
Hello World! qemu_leon3
Value: 6
```

0x6 != 0x60

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (nordicjm) [inline drivers/auxdisplay/auxdisplay_serlcd.c:38]: Something like
```suggestion
	struct __packed {
#ifdef CONFIG_LITTLE_ENDIAN
		uint8_t padding: 5;
		bool power: 1;
		bool cursor: 1;
		bool blinking: 1;
#else
		bool blinking: 1;
		bool cursor: 1;
		bool power: 1;
		uint8_t padding: 5;
#endif
	};
```

Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c]: Applied.

As a side note: This is the first time for me that *bit ordering* is referred to as endianess, which commonly refers to *byte ordering*.

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (nordicjm) [inline drivers/auxdisplay/auxdisplay_serlcd.c:173]:
```suggestion
		if (x < 0 || x >= columns) {
			return -EINVAL;
		}
		if (y < 0 || y >= rows) {
			return -EINVAL;
		}
```

**Review by nordicjm — COMMENTED** [review]
One final issue noticed

**Comment by janhenke** [comment]
All remaining comments addressed

**Review by nordicjm — DISMISSED** [review]

**Comment by janhenke** [comment]
Hi,
are there any remaining issues? I would like to get this merged.

**Comment by thedjnK** [comment]
Zephyr is in feature freeze mode for the 3.5 release, this can be merged after the release is complete (which should by this Friday)

**Comment by ShaharHD** [comment]
Any updates on this?

**Comment by ShaharHD** [comment]
@janhenke I've manully added it to my project, but can you also please include an example board (with I2C) to show how you actually define it inside the devicetree ?

Thanks :)

**Comment by thedjnK** [comment]
@fabiobaltieri please merge :)

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (fabiobaltieri) [inline drivers/auxdisplay/auxdisplay_serlcd.c:44]: you sure this is a good idea? normally I'd expect the driver file to `#define` the bit offset, keep whatever data you need in data and then prepare the command to send via i2c based on that. You could just do that, move these fields in `auxdisplay_serlcd_data`, saves you from having to manually deal with packed structures, unions etc...

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (fabiobaltieri) [inline drivers/auxdisplay/auxdisplay_serlcd.c:82]: add a define for `10`, give it a name like `SERLCD_COMMAND_DELAY`, then your comment become redundant as this would self-describe and you won't need it anymore

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (fabiobaltieri) [inline drivers/auxdisplay/auxdisplay_serlcd.c:98]: same as above, different define

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (fabiobaltieri) [inline drivers/auxdisplay/auxdisplay_serlcd.c:175]: do you need these aliases? why not just use x and y directly?

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (fabiobaltieri) [inline drivers/auxdisplay/auxdisplay_serlcd.c:236]: same comment as for the other sleep

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (fabiobaltieri) [inline drivers/auxdisplay/auxdisplay_serlcd.c:250]: add a define for the magic number

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (fabiobaltieri) [inline drivers/auxdisplay/auxdisplay_serlcd.c:260]: nit: usually good practice to keep the check next to line that sets the return value, drop this blank line

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (fabiobaltieri) [inline drivers/auxdisplay/auxdisplay_serlcd.c:272]: few magic here, add a define for 8 and 5?

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (fabiobaltieri) [inline drivers/auxdisplay/auxdisplay_serlcd.c:277]: use `sizeof(buffer)` instead of `1`, less likely to be copypasted wrong :-)

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (fabiobaltieri) [inline drivers/auxdisplay/auxdisplay_serlcd.c:316]: these could use few defines as well

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (fabiobaltieri) [inline drivers/auxdisplay/auxdisplay_serlcd.c:318]: drop the blank line

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (fabiobaltieri) [inline drivers/auxdisplay/auxdisplay_serlcd.c:392]: initializing this at build time causes the data structure to get a copy in flash rather than be zero-initialized automatically, it's not a big deal in this case as that's just 3 bytes but I think it's good practice to avoid that since you only need to initialize one bit from here and the rest is zero anyway

I suggest dropping the static initialization and just set the `power` bit at init time

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (fabiobaltieri) [inline drivers/auxdisplay/auxdisplay_serlcd.c:371]: bring the `{` on the line above

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (fabiobaltieri) [inline drivers/auxdisplay/auxdisplay_serlcd.c:381]: these could use a define (they show up in other spots anyway)

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (fabiobaltieri) [reply drivers/auxdisplay/auxdisplay_serlcd.c:44]: just to clarify where I stand on this: fine by me if @nordicjm is happy with it, but I don't think this is good practice and should be changed to work like most other drivers with defines for bit fields etc

Comment (thedjnK) [reply drivers/auxdisplay/auxdisplay_serlcd.c:44]: The define and bit offset suggestion sounds like a good improvement

**Comment by janhenke** [comment]
> @janhenke I've manully added it to my project, but can you also please include an example board (with I2C) to show how you actually define it inside the devicetree ?
> 
> Thanks :)

I not think it would be helpful to have a separate example project for every device driver, especially as this driver is pretty standard in terms of the devicetree.

But here you go:
```
&i2c1 {
	serlcd@72 {
		compatible = "sparkfun,serlcd";
		reg = <0x72>;
		columns = <16>;
		rows = <2>;
	};
};
```

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c:44]: The whole things is the result of previous review comments. As far as I am concerned, I am for the most simple solution.

Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c:44]: Do you have any driver which does the define/offset thing as a blueprint?

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c:82]: Done

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c:98]: Done

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c:175]: Mostly a "name the things so that it's purpose is clear" thing. If you prefer, I drop the alias.

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c:236]: Done

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c:250]: Done

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c:260]: IIRC the compliance workflow demanded this extra line, but I will split definition and declaration of rc, so it complies and is directly above.

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c:272]: Done

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c:277]: Done

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c:318]: Which blank line?

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c:316]: Done

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c:381]: Done

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c:371]: Done

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c:44]: I think I now have a non-packed version, albeit without offset but more explicit bit setting

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c:392]: Done

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (fabiobaltieri) [reply drivers/auxdisplay/auxdisplay_serlcd.c:341]: Oh, sorry misread the diff. You could use an intermediate variable for `SERLCD_COMMAND_WRITE_CUSTOM_CHAR + (character & 0x07)`? Would save this line from wrapping, the indentation is also a bit odd.

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (fabiobaltieri) [inline drivers/auxdisplay/auxdisplay_serlcd.c:34]: Keep the unit in the constant name, `SERLCD_COMMAND_DELAY_MS`, `SERLCD_SPECIAL_COMMAND_DELAY_MS` etc...

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (fabiobaltieri) [reply drivers/auxdisplay/auxdisplay_serlcd.c:341]: Also `~SERLCD_CUSTOM_CHAR_BITMASK` instead of 0x07?

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (fabiobaltieri) [reply drivers/auxdisplay/auxdisplay_serlcd.c]: Hey, yeah good stuff, that's what I meant, just use some defines for the bit definitoins now, ideally matching the name in the datasheet or manual or something.

**Comment by fabiobaltieri** [comment]
> ```
> &i2c1 {
> 	serlcd@72 {
> 		compatible = "sparkfun,serlcd";
> 		reg = <0x72>;
> 		columns = <16>;
> 		rows = <2>;
> 	};
> };
> ```

Since there's no upstream board with this you could add it in the `description` part of `dts/bindings/auxdisplay/sparkfun,serlcd.yaml` if you want, few other bindings do it.

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (fabiobaltieri) [inline drivers/auxdisplay/auxdisplay_serlcd.c:119]: I think it's a better idea to use a `uint8_t` here, the enum may change size depening on the compiler settings and you certainly don't want that to happen.

File: dts/bindings/auxdisplay/sparkfun,serlcd.yaml
Comment (fabiobaltieri) [inline dts/bindings/auxdisplay/sparkfun,serlcd.yaml:2]: Oh, just spotted this one, I think we need explicit approval for non Apache files, could you change it to `Apache-2.0`? We have a CI check for flagging these but it was broken for a while and just got fixed. :-(

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c:34]: Done

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c:119]: Done

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c:341]: Done

File: dts/bindings/auxdisplay/sparkfun,serlcd.yaml
Comment (janhenke) [reply dts/bindings/auxdisplay/sparkfun,serlcd.yaml:2]: Done

**Comment by nordicjm** [comment]
Compliance issues need fixed

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (nordicjm) [inline drivers/auxdisplay/auxdisplay_serlcd.c:52]: *a

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (nordicjm) [inline drivers/auxdisplay/auxdisplay_serlcd.c:62]: bitmask... for custom character?

**Review by nordicjm — COMMENTED** [review]
Minor changes

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c:52]: Done

File: drivers/auxdisplay/auxdisplay_serlcd.c
Comment (fabiobaltieri) [inline drivers/auxdisplay/auxdisplay_serlcd.c:430]: just

`static struct auxdisplay_serlcd_data auxdisplay_serlcd_data_##inst;                   \` (drop the ` = {}`)

Static variables are 0 initialized automatically.

Comment (janhenke) [reply drivers/auxdisplay/auxdisplay_serlcd.c]: Done

**Review by fabiobaltieri — APPROVED** [review]

**Review by nordicjm — APPROVED** [review]

**Pull request merged**
